doc/rfc001-sum-types.md

```diff
@@ -0,0 +1,846 @@
+# RFC 001: Extend Quint with Row-Polymorphic Sum Types
+
+**Table of Contents**
+
+- [tl;dr](#tldr)
+- [Motivation](#motivation)
+- [Context](#context)
+- [Statics](#statics-1)
+- [Dynamics](#dynamics-1)
+- [Concrete Syntax](#concrete-syntax)
+- [Additional consideration](#additional-consideration)
+
+## tl;dr
+
+This section gives a concise overview of the principle proposal. In depth
+discussion follows in the subsequent sections.
+
+### Concrete syntax
+
+Here I illustrate two attractive alternatives for each critical construct. See
+[Syntax](#syntax) for a discussion of motivations and trade offs.
+
+#### Declaration
+
+This group of alternatives is closer to Rust, but with significant variation for
+reasons discussed in [Syntax](#syntax).
+
+```quint
+type T =
+  | A(int)
+  | B(str)
+```
+
+Variants labels with no associated expression type are allowed, but these are only
+sugar for a label typed with the unit (i.e., the empty record):
+
+```quint
+type S = C
+```
+
+Is normalized as
+
+```quint
+type S = C({})
+```
+
+
+#### Injection
+
+I propose introducing a lexical rule to syntactically differentiate data
```

Comment:
What is the motivation for this lexical rule? As far as I understand, a type definition is introducing constructors for each option, similar to `pure def A(i: int) = ...`. Why would we need this lexical rule? We could say that is a **good practice** to lexically distinguish type names from operator names.

Reply:
I do mention this option. It’s just that I don’t think this is compatible with support for anonymous (not previously defined) sum types. If we require all sum types to be defined in advance, then we can um generate the constructors and don’t need the lexical rule. 

But as I note, I also think there is an inherent advantage to enabling readers to see what is a what is a sum type constructors just by looking at he code locally.